mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java

```diff
@@ -168,6 +132,129 @@ protected Call prepareCall(String url, String method, Map<String, String> additi
+                    TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            trustManagerFactory.init((KeyStore) null);
+            TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+
+            for (TrustManager trustManager : trustManagers) {
+                if (trustManager instanceof X509TrustManager) {
+                    return (X509TrustManager) trustManager;
+                }
+            }
+        } catch (Exception e) {
+            Log.d(TAG, "Getting default trust manager failed", e);
+        }
+
+        return null;
+    }
+
+    private static class DummyTrustManager implements X509TrustManager {
+        @Override
+        public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        }
+
+        @Override
+        public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        }
+
+        @Override
+        public X509Certificate[] getAcceptedIssuers() {
+            return new X509Certificate[0];
+        }
+    }
+
+    private static class ClientKeyManager implements X509KeyManager {
+        private final static String TAG = ClientKeyManager.class.getSimpleName();
+
+        private Context mContext;
+        private String mAlias;
+
+        public ClientKeyManager(Context context, String alias) {
+            mContext = context.getApplicationContext();
+            mAlias = alias;
+        }
+
+        @Override
+        public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
+            Log.d(TAG, "chooseClientAlias - alias: " + mAlias);
+            return mAlias;
+        }
+
+        @Override
+        public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+            Log.d(TAG, "chooseServerAlias");
+            return null;
+        }
+
+        @Override
+        public X509Certificate[] getCertificateChain(String alias) {
+            Log.d(TAG, "getCertificateChain", new Throwable());
+            try {
+                return KeyChain.getCertificateChain(mContext, alias);
+            } catch (KeyChainException | InterruptedException e) {
+                Log.e(TAG, "Failed loading certificate chain", e);
+                return null;
+            }
+        }
+
+        @Override
+        public String[] getClientAliases(String keyType, Principal[] issuers) {
+            Log.d(TAG, "getClientAliases");
+            return mAlias != null ? new String[] { mAlias } : null;
+        }
+
+        @Override
+        public String[] getServerAliases(String keyType, Principal[] issuers) {
+            Log.d(TAG, "getServerAliases");
+            return null;
+        }
+
+        @Override
+        public PrivateKey getPrivateKey(String alias) {
+            Log.d(TAG, "getPrivateKey", new Throwable());
+            try {
+                return KeyChain.getPrivateKey(mContext, mAlias);
```

Comment:
Are you sure that this method will not be called from main thread?
https://developer.android.com/reference/android/security/KeyChain.html

Reply:
As above.